tests/protractor/e2e/api/controllers/sms-gateway.spec.js

```diff
@@ -0,0 +1,347 @@
+      });
+
+      it('should return within a reasonable time', function() {
+        return saveWoMessages(...oneHundredWoMessages())
+          .then(() => {
+
+            const start = Date.now();
+            return postStatuses(...oneHundredUpdates())
+              .then(response => {
+                const end = Date.now();
+                const maxMillis = 10000; // TODO a _reasonable_ time would hopefully be less than 1s, not less than 10!
+                if(end > start + maxMillis) {
+                  const seconds = (end - start) / 1000;
+                  fail(`It took ${seconds}s to respond to the request.  The endpoint should respond within ${maxMillis}ms.`);
+                }
+                expect(response).toEqual({ messages:[] });
+              })
+
+              .then(() => getMessageStates())
+              .then(states => {
+                expect(states.length).toBe(100);
+
+                // expect: 1 state update per message
+                expect(
+                    states
+                      .map(s => s.states)
+                      .every(stateHistory => stateHistory.length === 1))
+                  .toBe(true);
+              });
+          });
+      });
+
+    });
+
+    it('should still save messages when a status update for an unknown message is received', function() {
+        return saveWoMessage('abc-123', 'hello again')
+          .then(() => expectMessageWithoutState('abc-123'))
+
+          .then(() => postStatuses(
+              { id:'abc-123', status:'SENT' },
+              { id:'def-456', status:'DELIVERED' }))
+          .then(expectResponse({ messages:[] }))
+          .then(() => expectMessageStates({ id:'abc-123', states:['sent'] }));
+    });
+
+    it('should still save messages when an unrecognised status update is received', function() {
+        return saveWoMessage('abc-123', 'hello again')
+          .then(() => expectMessageWithoutState('abc-123'))
+
+          .then(() => postStatus('abc-123', 'WTF'))
+          .then(expectResponse({ messages:[] }))
+          .then(() => expectMessageStates({ id:'abc-123', states:['unrecognised'] }));
+    });
+
+  });
+});
+
+function get() {
+  return utils.request({
+    path: '/api/sms',
+    method: 'GET',
+    headers: { 'Content-Type':'application/json' },
+  });
+}
+
+function post(body) {
+  return utils.request({
+    path: '/api/sms',
+    method: 'POST',
+    headers: { 'Content-Type':'application/json' },
+    body: body,
+  });
+}
+
+function postMessages(...messages) {
+  return post({ messages });
+}
+
+function expectResponse(expected) {
+  return response => expect(response).toEqual(expected);
+}
+
+function expectMessagesInDb(...expectedContents) {
+  expectedContents = JSON.stringify(expectedContents);
+
+  return new Promise((resolve, reject) => {
+    const endTime = Date.now() + 10000;
+
+    check();
+
+    function check() {
+      utils.db.query('medic-client/messages_by_contact_date', { reduce:false })
```

Comment:
This view is about to change significantly. @vimemo is working on a change that effectively removes the `value` so you won't be able to use `row.value.message` below. So you'll have to work out a different way to get the docs.

I feel like the best way may be to register a changes listener including docs before the endpoint is called and then assert the docs come through there within 10s. That removes the polling and will be independent of the view implementation.